From b99d1322f7ae3131cc54b8fa447e29e6fbc38059 Mon Sep 17 00:00:00 2001
From: SrLLamaDev <srllamadev@gmail.com>
Date: Sun, 13 Sep 2026 07:16:51 -0400
Subject: [PATCH 10/16] fix: harden general settings sanitization

update_general_settings() ran array_map() over an unvalidated value and
applied sanitize_text_field() to colors and URLs. Reject non-array
payloads, sanitize colors with sanitize_hex_color() and URLs with
esc_url_raw(), and sanitize remaining keys/values explicitly.
---
 .../inc/classes/rest-api/class-settings.php   | 41 +++++++++++++++++--
 1 file changed, 38 insertions(+), 3 deletions(-)

diff --git a/unlock-wordpress-plugin/inc/classes/rest-api/class-settings.php b/unlock-wordpress-plugin/inc/classes/rest-api/class-settings.php
index 770356b..bb3dd18 100644
--- a/unlock-wordpress-plugin/inc/classes/rest-api/class-settings.php
+++ b/unlock-wordpress-plugin/inc/classes/rest-api/class-settings.php
@@ -173,14 +173,49 @@ class Settings extends Rest_Base {
 	 *
 	 * @since 3.0.0
 	 *
-	 * @return bool
+	 * @return bool|\WP_Error
 	 */
 	public function update_general_settings( $data ) {
-		$data = array_map( 'sanitize_text_field', $data );
+		if ( ! is_array( $data ) ) {
+			return new \WP_Error(
+				'unlock_protocol_invalid_data',
+				__( 'Invalid settings payload.', 'unlock-protocol' ),
+				array( 'status' => WP_Http::BAD_REQUEST )
+			);
+		}
+
+		$color_fields = array(
+			'login_button_bg_color',
+			'login_button_text_color',
+			'checkout_button_bg_color',
+			'checkout_button_text_color',
+		);
+
+		$url_fields = array(
+			'checkout_url_base',
+			'locksmith_url_base',
+			'login_bg_image',
+			'checkout_bg_image',
+		);
+
+		$sanitized = array();
+
+		foreach ( $data as $key => $value ) {
+			$key = sanitize_key( $key );
+
+			if ( in_array( $key, $color_fields, true ) ) {
+				$color             = sanitize_hex_color( $value );
+				$sanitized[ $key ] = $color ? $color : '';
+			} elseif ( in_array( $key, $url_fields, true ) ) {
+				$sanitized[ $key ] = esc_url_raw( $value );
+			} else {
+				$sanitized[ $key ] = sanitize_text_field( $value );
+			}
+		}
 
 		$settings = get_option( 'unlock_protocol_settings', array() );
 
-		$settings['general'] = $data;
+		$settings['general'] = $sanitized;
 
 		$update = update_option( 'unlock_protocol_settings', $settings, false );
 
-- 
2.44.0.windows.1

